packages/pigeon/lib/ast.dart

```diff
@@ -78,29 +93,185 @@ class Method extends Node {
   }
 }
 
-/// Represents a collection of [Method]s that are hosted on a given [location].
-class Api extends Node {
+/// Represents a collection of [Method]s that are implemented on the platform
+/// side.
+class AstHostApi extends Api {
+  /// Parametric constructor for [AstHostApi].
+  AstHostApi({
+    required super.name,
+    required super.methods,
+    super.documentationComments = const <String>[],
+    this.dartHostTestHandler,
+  });
+
+  /// The name of the Dart test interface to generate to help with testing.
+  String? dartHostTestHandler;
+
+  @override
+  String toString() {
+    return '(HostApi name:$name methods:$methods documentationComments:$documentationComments dartHostTestHandler:$dartHostTestHandler)';
+  }
+}
+
+/// Represents a collection of [Method]s that are hosted on the Flutter side.
+class AstFlutterApi extends Api {
+  /// Parametric constructor for [AstFlutterApi].
+  AstFlutterApi({
+    required super.name,
+    required super.methods,
+    super.documentationComments = const <String>[],
+  });
+
+  @override
+  String toString() {
+    return '(FlutterApi name:$name methods:$methods documentationComments:$documentationComments)';
+  }
+}
+
+/// Represents an API that wraps a native class.
+class AstProxyApi extends Api {
+  /// Parametric constructor for [AstProxyApi].
+  AstProxyApi({
+    required super.name,
+    required super.methods,
+    super.documentationComments = const <String>[],
+    required this.constructors,
+    required this.fields,
+    this.superClassName,
+    this.interfacesNames = const <String>{},
+  });
+
+  /// List of constructors inside the API.
+  final List<Constructor> constructors;
+
+  /// List of fields inside the API.
+  List<Field> fields;
+
+  /// Name of the class this class considers the super class.
+  final String? superClassName;
+
+  /// Name of the classes this class considers to be implemented.
+  final Set<String> interfacesNames;
+
+  /// Methods implemented in the host platform language.
+  Iterable<Method> get hostMethods => methods.where(
+        (Method method) => method.location == ApiLocation.host,
+      );
+
+  /// Methods implemented in Flutter.
+  Iterable<Method> get flutterMethods => methods.where(
+        (Method method) => method.location == ApiLocation.flutter,
+      );
+
+  /// All fields that are attached.
+  ///
+  /// See [attached].
+  Iterable<Field> get attachedFields => fields.where(
+        (Field field) => field.isAttached,
+      );
+
+  /// All fields that are not attached.
+  ///
+  /// See [attached].
+  Iterable<Field> get unattachedFields => fields.where(
+        (Field field) => !field.isAttached,
+      );
+
+  @override
+  String toString() {
+    return '(ProxyApi name:$name methods:$methods documentationComments:$documentationComments superClassName:$superClassName interfacesNames:$interfacesNames)';
+  }
+}
+
+/// Represents a constructor for an API.
+class Constructor extends Node {
```

Comment:
Why does a constructor have to be a special type? It seems like all the properties are standard method properties.

Reply:
I think this is similar to our discussion about `parameter` vs `field`. They overlap but some of the parameters don't apply (e.g. `static`/`isAsynchronous`). However, if I consider the `Method` as the message call and not the AST class member, then I suppose they are essentially the same thing. I still need to distinguish the difference between `Constructors` for the Dart code generation, so I think I will keep the class and have it extend `Method`.